File: Greedy/MinCostToCutBoardIntoSq/minCost.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int minCost(int n, int m, vector<int>& x, vector<int>& y) {
        
        sort(x.begin(), x.end(), greater<int>());
        sort(y.begin(), y.end(), greater<int>());

        int horPiece = 1;
        int verPiece = 1;

        int horIn = 0;
        int verIn = 0;

        long long ans = 0;

        int nx = x.size();
        int ny = y.size();

        while (horIn < nx && verIn < ny) {

            if (x[horIn] > y[verIn]) {
                // x = vertical cut
                ans += (long long)horPiece * x[horIn];
                horIn++;
                verPiece++;
            }
            else {
                // y = horizontal cut
                ans += (long long)verPiece * y[verIn];
                verIn++;
                horPiece++;
            }
        }

        while (horIn < nx) {
            ans += (long long)horPiece * x[horIn];
            horIn++;
            verPiece++;
        }

        while (verIn < ny) {
            ans += (long long)verPiece * y[verIn];
            verIn++;
            horPiece++;
        }

        return ans;
    }
};

File: Greedy/MinCostToCutBoardIntoSq/minCost.py
class Solution:
    def minCost(self, n, m, x, y):
        x.sort(reverse=True)
        y.sort(reverse=True)

        arrX = x
        arrY = y

        horPiece = 1
        verPiece = 1
        horIn = 0
        verIn = 0
        ans = 0

        n = len(x)
        m = len(y)

        while horIn < n and verIn < m:
            if arrX[horIn] > arrY[verIn]:
                # x = vertical cut
                ans += horPiece * arrX[horIn]
                horIn += 1
                verPiece += 1
            else:
                # y = horizontal cut
                ans += verPiece * arrY[verIn]
                verIn += 1
                horPiece += 1

        while horIn < n:
            ans += horPiece * arrX[horIn]
            horIn += 1
            verPiece += 1

        while verIn < m:
            ans += verPiece * arrY[verIn]
            verIn += 1
            horPiece += 1

        return ans